Scalable consistent faders with themeable gradients, marker at unity, dbFS by default

michaelgregorius opened the pull request:
The changes mostly affect the mixer and the LMMS plugins Compressor, EQ and Delay.

# How does it look/feel?
A video is worth a thousand words:
https://github.com/LMMS/lmms/assets/9293269/d74e2070-2e54-4a4d-8351-34bce4a215b1

https://github.com/LMMS/lmms/assets/9293269/906307b3-045a-4cc1-b7f9-28243de960b6

# Detailed overview of the changes
The main changes are:
* Fader levels are rendered in code using a gradient instead of pixmaps. This ensures a consistent rendering. Code that used the pixmaps has been removed completely. The pixmaps have also been deleted.
* A marker at unity position is rendered, i.e. at position 1.0 in linear levels and 0 dbFS in dbFS scale. Rendering of the unity marker can be turned on and off via a style sheet property.
* It's now possible to specify the pixmap that's used to render the knob (again). So the "Crossover EQ" now renders with its own knob images again.
* Faders can be scaled/resized. The knob always renders along the mid line and the available space is used for the level indicators.
* The peak indicators do not use the three distinct colors anymore but instead use the color of the gradient at that position of the peak. This makes everything look "smooth".
* All faders now render in dbFS by default. This also improves the faders of the LMMS plugins Compressor, EQ and Delay.
* The default min and max values for the faders has been adjusted to -42 dbFS and 9 dbFS. This also improves Compressor, EQ and Delay. 
* The style sheet names for the colors have been renamed to something more neutral: peakOk, peakClip, peakWarn. The colors are used for the stops of the gradient.

The problem with the pixmap approach was that the pixmaps don't "know" how a fader instance is configured with regards to the minimum and maximum value. This means that the display can give quite a wrong impression. Artists that create pixmaps also don't know the technical details how to design them in a way that the code would render them correctly.

The minimum size of the faders is set to the previous size of the background pixmap.

Levels can technically still be rendered in linear but it should be considered to remove that feature.

Some code simplifications have also been applied: constructor delegation, remove constructors that are not needed anymore, remove `Fader::init`.

Some more technical details can be found in the commit messages of the individual commits.

Here's how larger faders render/look:
![FaderAdjustmentsLargerFadersExample2](https://github.com/LMMS/lmms/assets/9293269/8a88924f-4e2d-4cb8-b462-930a8000bb9e)

A clipping fader that shows the full scale of the gradient:
![FaderAdjustmentsClippingFaderFullScale](https://github.com/LMMS/lmms/assets/9293269/a1b6842f-f258-4d6b-8cd7-97040b25b96b)

Look of the LMMS Compressor:
![FaderAdjustmentsCompressor](https://github.com/LMMS/lmms/assets/9293269/6e9a503b-9de7-4f29-8e36-d1dded8025b9)

Look of the LMMS Equalizer:
![FaderAdjustmentsEqualizer](https://github.com/LMMS/lmms/assets/9293269/f12944a4-fc9d-444c-be1a-2829809c765a)

Look of the LMMS Delay:
![FaderAdjustmentsDelay](https://github.com/LMMS/lmms/assets/9293269/42c979e4-cedc-42ac-90fb-6b7e25a5466a)

Look of the LMMS Crossover EQ:
![FaderAdjustmentsCrossoverEQ](https://github.com/LMMS/lmms/assets/9293269/03b9a742-6b66-43f7-b48f-f29c5f2bc1ef)

michaelgregorius committed:
Render fader levels in code with a gradient

Render the fader level in code using a gradient instead of using pixmaps. The problem with the pixmaps is that they don't "know" how a fader instance is configured with regards to the minimum and maximum value. This means that the display can give quite a wrong impression.

The rendering of levels has been unified in the method `paintLevels`. It can render using dbFS and linear scale. The method `paintLinearLevels` has been removed completely, i.e. there's no more code that renders using pixmaps.

Much of the previous code relied on the size of the background image `fader_background.png`, e.g. the initialization of the size. For now the `Fader` widget is initially resized to the size of that background image as it is present in the default and classic theme (see `Fader::init`). All rendering uses the size of the widget itself to determine where to draw what. This means that the widget is prepared to be resizable.

The method `paintLevels` first renders the background of the level indicators and uses these as clipping paths for all other rendering operations, e.g. for the rendering of the levels themselves. Levels are rendered using a gradient which is defined with the following stops:
* Two stops for the ok levels.
* One stop for warning levels.
* One stop for clipping levels.

Peak indicators do not use the three distinct colors anymore but instead use the color of the gradient at that position of the peak. This makes everything look "smooth".

The code now also renders a marker at unity position, i.e. at position 1.0 in linear levels and 0 dbFS in dbFS scale.

The painting code makes lots of use of the class `PaintHelper`. This class is configured with a minimum and maximum value and can then return linear factors for given values. There are two supported modes:
* Map min to 0 and max to 1
* Map min to 1 and max to 0

It can also compute rectangles that correspond to a given value. These methods can be given rectangles that are supposed to represent the span from min to max. The returned result is then a rectangle that fills the lower part of the source rectangle according to the given value with regards to min and max (`getMeterRect`). Another method returns a rectangle of height 1 which lies inside the given source rectangle at the corresponding level (`getPersistentPeakRect`).

The method `paintLevels` uses a mapping function to map the amplitude values (current peak value, persistent peak, etc.) to the display values. There's one mapper that keeps the original value and it is used to display everything in a linear scale. Another mapper maps everything to dbFS and uses these values as display everything in a dbFS scale. The following values must be mapped for the left and right channel to make this work:
* Min and max display values (min and max peak values)
* The current peak value
* The persistent peak value
* The value for unity, i.e. 1.0 in linear levels and 0 dbFS in dbFS scale.

Remove the method `calculateDisplayPeak` which was used in the old method to render linear levels.

`Fader::setPeak` now uses `std::clamp` instead of doing "manual" comparisons.

The LMMS plugins Compressor, EQ and Delay are still configured to use linear displays. It should be considered to switch them to dbFS/logarithmic displays as well and to remove the code that renders linearly.

michaelgregorius committed:
Remove unused pixmaps from `Fader`

Remove the now unused pixmaps for the background and the LEDs from the `Fader` class and remove the files from the default and classic theme directories.

michaelgregorius committed:
Rename peak properties and use them to render levels

Rename the peak properties as follows:
* peakGreen -> peakOk
* peakRed -> peakClip
* peakYellow -> peakWarn

The reasoning is that a style might for example use a different color than green to indicate levels that are ok.

Use the properties to initialize the gradient that is used to render the levels.

Initialize the properties to the colors of the current default theme so that it's not mandatory to set them in a style sheet. Up until now they have all been initialized as black.

michaelgregorius committed:
Always render the knob in the middle of the fader

Render the knob in the middle of the fader regardless of the width. The previous implementation was dependent on the fader pixmap having a matching width because it always rendered at x=0.

michaelgregorius committed:
Set size policy of fader to minimum expanding

Set the size policy of the fader to minimum expanding in both directions. This will make the fader grow in layouts if there is space.

michaelgregorius committed:
Default dbFS levels and better peak values

Default to dbFS levels for all faders and set some better minimum and maximum peak values.

michaelgregorius committed:
Fix faders of Crossover EQ

Fix the faders of the Crossover EQ which were initialized and rendered much too wide and with a line at unity. The large width also resulted in the knobs being rendered outside of view.

Resize the fader to the minimum size so that it is constructed at a sane default.

Introduce a property that allows to control if the unity line is rendered. The property is available in style sheets and defaults to the unity lines being rendered. Adjust the paint code to evaluate the property.

Initialize the faders of the Crossover EQ such that the unity line is not drawn.

michaelgregorius committed:
Remove EqFader constructor with pixmaps

Remove the constructor of `EqFader` that takes the pixmaps to the fader background, leds and knob. The background and leds pixmaps are not used by the base class `Fader` for rendering anymore to make the `Fader` resizable. A pixmap is still used to render the knob but the constructor that takes the knob as an argument does not do anything meaningful with it, i.e. all faders are rendered with the default knob anyway.

Remove the resources for the fader background, leds and knob as they are not used and the knob was the same image as the default knob anyway.

Remove the static pixmaps from the constructor of `EqControlsDialog`. Switch the instantiations of the EQ's faders to use the remaining constructor of `EqFader`. This constructor sets a different fixed size of (23, 116) compared to the removed constructor which set a size of (23, 80). Therefore all faders that used the removed constructor are now set explicitly to a fixed size of (23, 80).

The constructor that's now used also calls a different base constructor than the removed one. The difference between the two base constructors of `Fader` is that one of them sets the member `m_conversionFactor` to 100.0 whereas the other one keeps the default of 1.0. The adjusted faders in `EqControlsDialog` are thus now constructed with the conversion factor set to 100. However, all of them already call `setDisplayConversion` with `false` after construction which results in the conversion factor being reset to 1.0. So the result should be the same as before the changes.

michaelgregorius committed:
Remove background and LEDs pixmap from Fader constructor

Remove the parameters for the background and LEDs pixmap from the second `Fader` constructor. Make the knob pixmap parameter in the constructor a const reference. Assign the reference to the knob pixmap of the `Fader` itself. This enables clients to use their own fader knobs as is the case with the Crossover EQ. The EQ now renders using it's own knobs again.

Make the second constructor delegate to the first one. This will additionally set the conversion factor to 100 but this is not a problem with the current code because the only user of the second constructor, the Crossover EQ, already calls `setDisplayConversion` with the parameter set to `false`, hence reinstating a conversion factor of 1.

Remove the resources for the background and LEDs from the Crossover EQ as they are not used anymore. Remove the three QPixmap members from `CrossoverEQControlDialog` as they are not needed. The background and LEDs are not used anyway and the knob is passed in as a constant reference which is copied. Hence we can use a local variable in the constructor of `CrossoverEQControlDialog`.

michaelgregorius committed:
Remove the init method from Fader

Remove the `init` method from `Fader` as it is not needed anymore due to the constructor delegation. Tidy up the parameter lists and use of spaces in the constructor.

michaelgregorius committed:
Merge remote-tracking branch 'upstream/master' into FaderInCodeWithGradients

Conflicts:
* include/Fader.h
* src/gui/widgets/Fader.cpp

he29-net commented:
Some small issues at UHD resolution (2x scale):
 - the Mixer is a bit oversized: each fader is a bit wider (possibly related to the increased padding in LCDs?) and taller (probably because of increased padding between SEND/knob/arrows);
 - some of the longer labels are cropped at the beginning;
 - LED toggles seem to be drawn at 2x size but within 1x boundary.
![screen](https://github.com/LMMS/lmms/assets/20932751/2d94dc2b-a795-4c20-97dc-a3161898f43d)

I'm not sure about the smooth gradients: the sharp contrast between green and yellow was very easy to notice at a glance, drawing attention to channels that may be approaching clipping. Not so much with the gradient. Maybe the gradient stops could be doubled, for example using thresholds like peakWarnLow = 0.75, peakWarnHigh = 0.8, resulting in a sharper change and areas that are fully green, yellow, and red? (The number is completely random, no idea what thresholds are currently used). But it may be just a personal preference.

Great work overall; a nice step towards fully scalable UI.

he29-net commented:
One even smaller issue: the unity line overlaps with the borders, resulting in bright pixels where they cross:
![screen](https://github.com/LMMS/lmms/assets/20932751/0e4ddb52-ec15-4848-9e2c-a9a25a4d6f9a)

michaelgregorius commented:
Thanks for checking @he29-net!

> Some small issues at UHD resolution (2x scale):
> 
>     * the Mixer is a bit oversized: each fader is a bit wider (possibly related to the increased padding in LCDs?) and taller (probably because of increased padding between SEND/knob/arrows);
> 
>     * some of the longer labels are cropped at the beginning;
> 
>     * LED toggles seem to be drawn at 2x size but within 1x boundary.

I think all these issues are related to the layout changes that have been made with pull request #6591. So I propose to create a new issue for these so that we do not mix topics (mixer channels with layouts vs. fader changes).

> I'm not sure about the smooth gradients: the sharp contrast between green and yellow was very easy to notice at a glance, drawing attention to channels that may be approaching clipping. Not so much with the gradient. Maybe the gradient stops could be doubled, for example using thresholds like peakWarnLow = 0.75, peakWarnHigh = 0.8, resulting in a sharper change and areas that are fully green, yellow, and red? (The number is completely random, no idea what thresholds are currently used). But it may be just a personal preference.

We can also double the yellow area , similar to how it's done for the green area (see [here](https://github.com/LMMS/lmms/blob/0126c027be908c219ea8f246c4c798a79d11a3ea/src/gui/widgets/Fader.cpp#L401) in the code).

With the current code everything below -12 dbFS is fully green. Then it becomes yellow as it approaches unity and from there on it goes into red. So as part of the fix we could also increase the dbFS value of the last fully green signal, e.g. -6 dbFS or -3 dbFS.

Any preferences with regards to the transitions? What should be the last fully green dbFS value? At how many dbFS should we approach the full on warning area? How far should it exceed? At what dbFS value are we at full clipping?

> Great work overall; a nice step towards fully scalable UI.

Thanks! :smiley:

michaelgregorius commented:
> One even smaller issue: the unity line overlaps with the borders, resulting in bright pixels where they cross: 
> [snip]

I think this might be caused by the order of rendering. Currently I render as follows:
* Draw the rounded background (and use it to clip the remaining draw calls)
* Draw the levels
* Draw the unity lines
* Finally do some white-ish overdraw of the background's outline to approach the look of the original pixmap.

I think it might help if I draw the unity lines as the very last step.

michaelgregorius committed:
Draw unity lines last

Draw the unity lines last so that they are not affected by the white-ish overdraw in line 423 (after the changes).

michaelgregorius commented:
With commit 045a2cd5c3f the unity lines are now drawn last.

michaelgregorius committed:
Introduce range with solid warn color

Introduce a second point in the gradient for the warn colors so that we get a certain range with the full/solid warn color.

The colors are distributed as follows now. The solid ok range goes from -inf dbFS to -12 dbFS. The warn range goes from -6 dbFS to 0 dbFS. In between the colors are interpolated. Values above 0 dbFS interpolate from the warn color to the clip color.

This is now quite similar to the previous implementation.

# Analysis of the previous pixmap implementation
The pixmap implementation used pixmaps with a height of 116 pixels to map 51 dbFS (-42 dbFS to 9 dbFS) across the whole height. The pixels of the LED pixmap were distributed as follows along the Y-axis:
* Margin: 4
* Red: 18
* Yellow: 14
* Green: 76
* Margin: 4

Due to the margins the actual red, yellow and green areas only represent a range of (1 - (4+4) / 116) * 51 ~ 47,48 dbFS. This range is distributed as follows across the colors:
Red: 7.91 dbFS
Yellow: 6.16 dbFS
Green: 33.41 dbFS

The borders between the colors are located along the following dbFS values:
* Red/yellow: 9 - (4 + 18) / 116 * 51 dbFS ~ -0.67 dbFS
* Yellow/green: 9 - (4 + 18 + 14) / 116 * 51 dbFS ~ -6.83 dbFS
* The green marker is rendered for values above -40.24 dbFS.

michaelgregorius commented:
@he29-net, with commit a644bdcd4f8 there is now a warn range with a solid yellow color. It goes from -6 dbFS to 0 dfBS. I have analyzed the previous implementation (see below) and have chosen the values such that they look very similar.

Here's how it looks:
![Minus12ToMinus6](https://github.com/LMMS/lmms/assets/9293269/31c38cfb-1728-4474-b1b8-f92cedbd972b)

# Analysis of the previous pixmap implementation
The pixmap implementation used pixmaps with a height of 116 pixels to map 51 dbFS (-42 dbFS to 9 dbFS) across the whole height. The pixels of the LED pixmap were distributed as follows along the Y-axis:
* Margin: 4
* Red: 18
* Yellow: 14
* Green: 76
* Margin: 4

Due to the margins the actual red, yellow and green areas only represent a range of (1 - (4+4) / 116) * 51 ~ 47,48 dbFS. This range is distributed as follows across the colors:
Red: 7.91 dbFS
Yellow: 6.16 dbFS
Green: 33.41 dbFS

The borders between the colors are located along the following dbFS values:
* Red/yellow: 9 - (4 + 18) / 116 * 51 dbFS ~ -0.67 dbFS
* Yellow/green: 9 - (4 + 18 + 14) / 116 * 51 dbFS ~ -6.83 dbFS
* The green marker is rendered for values above -40.24 dbFS.

michaelgregorius committed:
Remove unused method Fader::clips

michaelgregorius committed:
Fader: Correctly render arbitrary ranges

Adjust the `Fader` so that it can correctly render arbitrary ranges of min and max peak values, e.g. that it would render a display range of [-12 dbFS, -42 dbFS] correctly.

Until now the gradient was defined to start at the top of the levels rectangle and end at the bottom. As a result the top was always rendered in the "clip" color and the bottom in the "ok" color. However, this is wrong, e.g. if we configure the `Fader` with a max value of -12 dbFS and a min value of -42 dbFS. In that case the whole range of the fader should be rendered with the "ok" color.

The fix is to compute the correct window coordinates of the start and end point of gradient using from the "window" of values that the `Fader` displays and then to map the in-between colors accordingly. See the added comments in the code for more details.

Add the templated helper class `LinearMap` to `lmms_math.h`. The class defines a linear function/map which is initialized using two points. With the `map` function it is then possible to evaluate arbitrary X-coordinates.

michaelgregorius committed:
Fix whitespace

michaelgregorius committed:
Remove unused methods in PaintHelper

Remove the now unused mapping methods from `PaintHelper`. Their functionality has been replaced with the usage of `LinearMap` in the code.

michaelgregorius committed:
Fix some builds

Include `cassert` for some  builds that otherwise fail.

michaelgregorius commented:
Hi @LMMS/developers,

I consider this pull request done now and would like to request a review so that the changes can be merged. Please note that this PR is not just a cosmetic change but that it fixes fundamental functionality of the `Fader` class. Therefore I propose to include this PR in the next 1.3 release.

The fix can be demonstrated by configuring the `Fader` with a minimum peak value of -42 dbFS and a maximum peak value of -12 dbFS. In this case the correct rendering is an all green line if the fader is maxed out. This in fact now happens with the PR:

https://github.com/LMMS/lmms/assets/9293269/f1ee2caf-9e21-4fe8-91f6-f170c0c6a853

And here's how the setup with a range of [-42, -12] dbFS renders in the current master:

https://github.com/LMMS/lmms/assets/9293269/524c573d-efc4-477c-a774-446c37ca88b9

The master just renders the whole pixmap which makes it look as it there's some heavy clipping going on when in fact it isn't.

Please refer to the original PR message for all other additional improvements.

he29-net commented:
> with commit https://github.com/LMMS/lmms/commit/a644bdcd4f8db2a515a039a15fc15c068b91c6b4 there is now a warn range with a solid yellow color.

Great, subjectively I think it looks better now. Maybe the orange just above the unity line could more aggressively inform the user that "hey, danger, we are clipping now", but I think the addition of the unity line itself takes care about that. Combined with the colored peak markers I think it is sufficient.

Speaking of the unity line, https://github.com/LMMS/lmms/commit/045a2cd5c3f9956b912100dfd5aeb0ef2ee5cf87 did not really change anything for me. This is how it looks in the mixer:
![screen](https://github.com/LMMS/lmms/assets/20932751/222d8edb-04e9-45d7-b5f1-604352b01913)

It's not as pronounced on the lighter background compared to EQ, but the bright dots are still there. I think the issue is simply the alpha of 127 in `unityMarkerColor`, which allows it to mix with the dark (but not black) border under it. So maybe the line should be made completely opaque, or drawn one pixel shorter on each side so there is no overlap?

As a side note, I almost complained about uneven alignment on the red faders; turns out it is real, but it's caused by the physical distance of red and green subpixels on my LCD... :) So not a code issue.
![IMG_5748_crop](https://github.com/LMMS/lmms/assets/20932751/ab6fd6cb-9427-4829-8147-9632f9dfad1c)
(8 white pixels were added in editor)

michaelgregorius committed:
Opaque unity marker with styling option

Make the unity marker opaque by default and enable to style it with the style sheets. None of the two style sheets uses this option though.

michaelgregorius commented:
Hi @he29-net,

with commit 99113c3b309 the unity marker now defaults to an opaque color. It is now also possible to style this color via the style sheet.

With regards to warning the users about clipping: I think in some situations, especially with regards to mixing in digital, it might be okay for one or more channels to "clip", for example if they get routed to another channel which attenuates the mix again. On the master channel on the other hand it's interesting to get warned about clipping though. So it might all be relative anyway.

he29-net commented:
After 99113c3 the unity line is now pixel-perfect. :+1: 

Maybe (63, 63, 63, 255) would be a better choice for the default though, if we are not updating the styles? It would more or less match the previous appearance, i.e. it would blend better with the dark background.

(Sorry for constantly pointing out tiny graphical issues.. ^^; This is why I prefer working on embedded systems: there is usually no graphical interface the user can complain about, haha..)

michaelgregorius committed:
Darker default color for the unity line

michaelgregorius committed:
Remove TODO

michaelgregorius commented:
> After [99113c3](https://github.com/LMMS/lmms/commit/99113c3b309cbd10900be8ab7daab1b86be74c6b) the unity line is now pixel-perfect. 👍

Ok, great. :slightly_smiling_face: By the way, I think that screenshot was the ultimate (sub) pixel peeping. :smile: 

> Maybe (63, 63, 63, 255) would be a better choice for the default though, if we are not updating the styles? It would more or less match the previous appearance, i.e. it would blend better with the dark background.

Done with commit a44b654f31a.

> (Sorry for constantly pointing out tiny graphical issues.. ^^; This is why I prefer working on embedded systems: there is usually no graphical interface the user can complain about, haha..)

No problem! Next up then: LMMS on an ESP32. :smile:

RoxasKH commented:
> I'm not sure about the smooth gradients: the sharp contrast between green and yellow was very easy to notice at a glance, drawing attention to channels that may be approaching clipping. Not so much with the gradient. Maybe the gradient stops could be doubled, for example using thresholds like peakWarnLow = 0.75, peakWarnHigh = 0.8, resulting in a sharper change and areas that are fully green, yellow, and red? (The number is completely random, no idea what thresholds are currently used). But it may be just a personal preference.
> One even smaller issue: the unity line overlaps with the borders, resulting in bright pixels where they cross

I fully agree with this: i feel like the full gradient doesn't really fit the current LMMS flat style and it also makes it a lot more difficult to spot clipping.
The marker is there to help with it, but it's not really pleasant to look at, as it's a sharp gray line over the volume levels and the fact it's larger than the volume levels doesn't help.

Taking a glance at other daws which uses a full-bar approach at showing volume levels like LMMS (instead of a steps one), they all use a mixed approach, with a gradient on the green/yellow portion, but keeping a straight and sudden passage to red when it clips.

While for the marker, they all have it either outside of the mixer bars, or it just stays quietly behind the levels (as it's more aestethic than anything else at that point considering the red easily marks the clipping levels).

![Mixers](https://github.com/LMMS/lmms/assets/40141826/c78c8be6-b95e-4959-853f-19a8385e8741)

In a theme i was working on, which is a recolor that kind of follows the color scheme used in the single window mockup presented in #1911 , i did something similar too, although the mockup didn't show how clipping would have behaved, i had the green gradient but kept the yellow and red sections flat, and also added background markers in a similar way to what FL Studio does.
(In this specific example i also made the volume bars larger but that's not something to care about here).

![immagine](https://github.com/LMMS/lmms/assets/40141826/2224d267-6936-4b62-b5f4-49b1e866ec11)

I'm not saying that's the exact path to follow, as i showed lots of daws have a linear gradient between green and yellow.
As said by others, too, i appreciate the work to a better scalable UI, but personally i'd still rather keep at least the clipping to be visible through a clear separation of a red portion of the levels, instead of using a full gradient and overlap an indicator on it.

At least this is my opinion

michaelgregorius committed:
Move code

Move the computation of most mapped values at the top right after the definition of the mapper so that it is readily available in all phases of the painting code.

michaelgregorius committed:
Render unity lines in background

Render the unity lines before rendering the levels so that they get overdrawn and do not stick out when they are crossed.

michaelgregorius committed:
Don't draw transparent white lines anymore

Don't draw the transparent white lines anymore as they were mostly clipped anyway and only create "smudge".

michaelgregorius commented:
Hi @RoxasKH, thanks for your feedback! I really appreciate it. With commit 5627384833a I have move the unity marker in the background so that it gets overdrawn when the levels surpass it. It looks better indeed.

I think having the markers outside might clash with the option for the users to color the mixer channels. In some cases the markers might be barely visible. One option here might be to render them on top of some slightly transparent black rectangle, so that the colors can shine through but the markers (and later values) would still be readable.

When I find the time I will experiment with drawing the clip color as a solid rectangle which starts right at 0 dbFS.

michaelgregorius committed:
Full on clip color at unity

Adjust the gradient so that the full on clip color shows starting at unity. There is only a very short transistion from the end of warning to clipping making it look like a solid color "standing" on top of a gradient.

michaelgregorius committed:
Fix discrepancy between levels and unity markers

This commit removes the helper class `PaintHelper` and now uses two lambdas to compute the rectangles for the peak indicators and levels. It uses the linear map which maps the peak values (in dbFS or linear) to window coordinates of the widget.

The change fixes a discrepancy in the following implementation for which the full on clip rectangle rendered slightly below the unity marker.

michaelgregorius commented:
@RoxasKH, I have added some commits which render the clipping area as a flat/solid region. The full scale currently looks as follows:

![LevelsWithSolidClipRegion](https://github.com/LMMS/lmms/assets/9293269/0360b9f1-083e-4fb9-98eb-3e266dea26dd)

The change also showed that there was a problem with the unity marker. This was fixed by replacing the helper class `PaintHelper` with another mechanism.

Here's a test file that can be used to test the fader. It consists of a sine that's played at unity:
[MixerDevelopmentSineAtZero.zip](https://github.com/LMMS/lmms/files/13866521/MixerDevelopmentSineAtZero.zip)

michaelgregorius committed:
Merge remote-tracking branch 'upstream/master' into FaderInCodeWithGradients

Conflicts:
* include/lmms_math.h

michaelgregorius commented:
@PhysSong, do you also want to review this PR? I would greatly appreciate if it would be merged it as it brings several improvements.

Rossmaxx commented:
Just a doubt. There seems to be a seperate `EqFader` class which is inherited from the `Fader` class and from what I've heard, it's difference is only the fact that EqFader renders in dbFs while the mixer fader renders in linear levels. So my question is, can you use the regular fader for eq and get rid of eq fader after this pr?

michaelgregorius commented:
> Just a doubt. There seems to be a seperate `EqFader` class which is inherited from the `Fader` class and from what I've heard, it's difference is only the fact that EqFader renders in dbFs while the mixer fader renders in linear levels. So my question is, can you use the regular fader for eq and get rid of eq fader after this pr?

Hi @Rossmaxx, I guess the `Fader` class will then need to be extended with another property which governs if the input data is linear or in dbFS. Currently the `Fader` class assumes linear input data which might also be a problem if the `EqFader` (and therefore the `Fader`) is fed levels which are already in dbFS.

curlymorphic commented:
I can sure try, although it was a decade ago I wrote this code. Can you confirm what line you are referring to please?

michaelgregorius commented:
Thanks for the quick response, @curlymorphic!

It's mainly about the lines [299](https://github.com/LMMS/lmms/blob/360254fd81cda14cbcec3fff3dff5456005ebfc8/plugins/Eq/EqEffect.cpp#L299) and [304](https://github.com/LMMS/lmms/blob/360254fd81cda14cbcec3fff3dff5456005ebfc8/plugins/Eq/EqEffect.cpp#L304) in the method [EqEffect::peakBand](https://github.com/LMMS/lmms/blob/360254fd81cda14cbcec3fff3dff5456005ebfc8/plugins/Eq/EqEffect.cpp#L290).

curlymorphic commented:
I can sure help,
 line 299, ````h = 20 * ( log10( *b / fft->getEnergy() ) );```` converts the linear power as return by an fft to dBfs.
line 300 simply clamps this value, so any value lower than -60 = -60,  so the display does not respond to low level signals. This would have been common on old analog mixing desks from the 1960's to maybe 1990's, as -60dB would have been considered the nosie floor.

I am having to think about line 304, and git blame shows this has been changed since my commit, but it looks like it is simply converting the dbfs value to a normalised 0.0f < h < 1.0f range for display.

The conversions are required as dB scales are non linear to represent how the ear responds to the change in power levels.

michaelgregorius commented:
Thanks for the further input @curlymorphic!

I will then try to adjust the `Fader` class so that it can be configured to also take dbFS values as input and render them.

Doing so will enable the following changes to [EqEffect::peakBand](https://github.com/LMMS/lmms/blob/360254fd81cda14cbcec3fff3dff5456005ebfc8/plugins/Eq/EqEffect.cpp#L290):
* It can return the raw dbFS value instead of some mapped value.
* It will also make sense to reduce the minimum decibel value to something like -144 dbFS because that seems to be the noise floor for `float` values. After all we are working in digital and not with noisy consoles from the 60ies. :wink: How these values are displayed, i.e. their minimum and maximum range, should be implemented in the presentation layer and not in the "business"/core layer.

I am also wondering about several things:
* Is the current mapping correct in the first place? I am asking because it seems that it will map a peak of 40 dbFS to unity instead of 0 dbFS: `(40 + 60) / 100 = 1`.
* Why is the maximum peak of a single bin taken as the peak value? I'd argue that the other bins will in many cases have an effect as well. If I'd take the IFFT of the bins in the frequency range then the individual waves could interfere constructively or destructively due to phase, etc.

curlymorphic commented:
I was looking yesterday for an old post where Diizy, the lead dev at the time was insistent as to how this should be displayed. I seem to remember stating that this behavior is not the same as all the other DAW's. I agree the range is not what I would expect. This could well be the time for change.

I have no clue as to why this line of code was ever to use (40+60) / 100, but I would assume at some point someone didn't like the original implementation. The original code that I found using git blame was ````return (peak+100)/100;````

 Why is the maximum peak of a single bin taken as the peak value?
 Because this was a simple visualisation, that would give values that matched the large fft display area. You are correct that multiple bins within the range could have better analysis, but you may wish to deal with that as a separate issue, as it may well be outside the scope of this PR.

As a bit of history, The EQ plugin was my first ever released module, and you may well find many areas that need improvement. I looked at the images in the original PR and it looks very different today, a credit to all the developers that have done revisions since. Thank you for continuing the effort.

michaelgregorius committed:
Fix fader display for Equalizer shelves and peaks

The peak values for the shelves and peaks of the Equalizer plugin are computed in `EqEffect::peakBand`. The previous implementation evaluated the bins of the corresponding frequency spectrum and determined the "loudest" one. The value of this bin was then converted to dbFS and mapped to the interval [0, inf[ where all values less or equal to -60 dbFS were mapped to 0 and a value of 40 dbFS was mapped to 1. So effectively everything was mapped somewhere into [0, 1] yet in a quite "distorted" way because a signal of 40 dbFS resulted in being displayed as unity in the fader.

This commit directly returns the value of the maximum bin, i.e. it does not map first to dbFS and then linearize the result anymore. This should work because the `Fader` class assumes a "linear" input signal and if the value of the bin was previously mapped to dbFS it should have some "linear" character. Please note that this is still somewhat of a "proxy" value because ideally the summed amplitude of all relevant bins in the frequency range would be shown and not just the "loudest" one.

## Other changes
Rename `peakBand` to `linearPeakBand` to make more clear that a linear value is returned.

Handle a potential division by zero by checking the value of `fft->getEnergy()` before using it.

Index into `fft->m_bands` so that no parallel incrementing of the pointer is needed. This also enables the removal of the local variable `b`.

michaelgregorius commented:
I have solved this differently now. Instead of extending the `Fader` for this single case I decided to remove the mapping to [0, 1] (or technically rather [0, inf[) that's done at the end and to convert the dbFS values to linear values because that's what the `Fader` expects with regards to the peaks. Then I realized that the code would effectively do `linear -> dbFS -> linear` if I did this and that I can simply use the maximum bin as calculated in the `for` loop.

The corresponding changes are introduced via commit 449ab02a which also introduces the handling of a potential division by zero. It also removes the clamping to -60 dbFS so that the fader can decide how to present the raw data.

Thanks for introducing the Equalizer in the first place, @curlymorphic! :slightly_smiling_face:

michaelgregorius commented:
I consider this pull request ready for review and merge now. There are several other things that could be done in separate PRs once this one is merged:
* Render and evaluate the fader knob on a dbFS scale instead of a linear one. The models should be kept linear though because they can be automated and switching them to dbFS values would likely introduce too much data upgrade efforts for save files. A linear model with the interval [0, 2] also enables real silence whereas a dbFS model would have to map -∞ somehow.
* Make the mixer vertically resizable so that additional available space could be use by the faders.
* Unification of `Fader` and `EqFader` (if it is really wanted)
* More accurate computation of the Equalizer peak levels (if it is efficiently possible in FFT space)

RoxasKH commented:
> @RoxasKH, I have added some commits which render the clipping area as a flat/solid region. The full scale currently looks as follows:
> 
> The change also showed that there was a problem with the unity marker. This was fixed by replacing the helper class `PaintHelper` with another mechanism.
> 
> Here's a test file that can be used to test the fader. It consists of a sine that's played at unity: [MixerDevelopmentSineAtZero.zip](https://github.com/LMMS/lmms/files/13866521/MixerDevelopmentSineAtZero.zip)

I know i'm late but I'm in favour of this cause LMMS needs everything it can have to be dinamically rendered, so i'll add a bit from a design perspective.

Here's a 1:1 view against the current faders

![immagine](https://github.com/LMMS/lmms/assets/40141826/20e4c4d4-2f7b-4965-a8e6-ea945dc88a16)

Now, aside from the sizes of the mixer buses itself

1. We're losing both the slight outline and the curvature at the ends of the faders, which made it more polished than the abrupt cut which is now present to m

![immagine](https://github.com/LMMS/lmms/assets/40141826/cb362a1a-df5e-41be-992b-b31003d2b7cf)

2. The volume levels dimensions changed? Now bars have a width of 6px, while the old one were of 7. This is inherently bad, but while there are other daws with thin levels, they do not have the volume fader overlayed on the levels themselves, so i believe having them larger is actually better.

![immagine](https://github.com/LMMS/lmms/assets/40141826/211d389f-683c-4d04-a69d-f50742a4c92a)

3. The bars is slightly larger than the fader so it's 7 px. This creates a weird effect where the bar seems to be getting thinner when it peaks, as the light color of the bar kind of merges with the yellow part of the volume. Now there's more than 1 single possible fix; FL Studio which is my reference cause it's the only one having bars on the volume levels, keep them thinner than the level, so that the bar is actually visible only when it's not hovered: at the end of the day once the bar gets covered the red color itself will tell the producer about the clipping.

![immagine](https://github.com/LMMS/lmms/assets/40141826/8647c048-f41a-456f-bc88-0df58c6baf75)

On active faders the color of the bar makes it merge with the background as well, so it might be worth changing color? Not sure actually, once made them thinner there should be a better separation.
Actually, by bringing back the volume levels to 7px wide they should cover the line by themselves with no need to make them thinner i guess.

![immagine](https://github.com/LMMS/lmms/assets/40141826/2a53071e-de43-4b68-bf96-3ffda296d177)

Obviously this are probably all minor design things, but i believe they should be addressed when this will substitute the current faders.

michaelgregorius committed:
Improve the rendering of the levels

The levels rendering now more explicitly distinguished between the rendering of the level outline/border and the level meters. The level rectangles are "inset" with regards to the borders so that there is a margin between the level borders and the meter readings. This margin is now also applied to the top and bottom of the levels. Levels are now also rendered as rounded rectangles similar to the level borders.

Only render the levels and peaks if their values are greater than the minimum level.

Make the radius of the rounded rectangles more pronounced by increasing its value from 1 to 2.

Decrease the margins so that the level readings become wider, i.e. so that they are rendered with more pixels.

Add the lambda `computeLevelMarkerRect` so that the rendering of the level markers is more decoupled from the rendering of the peak markers.

michaelgregorius commented:
Thanks for your feedback,  @RoxasKH!

Most of your points should have been addressed by commit b4bfc7174fa:
* It introduces a margin in all directions between the level borders and the level meters.
* Levels are now also rendered as rounded rectangles similar to the level borders.
* The radius of the rounded rectangles was increased to make it more pronounced.
* The margins have been decreased so that the level readings become wider.
* The unity indicator is now rendered with the same width as the level meters so that they become fully covered by them.

Other improvements:    
* Only render the levels and peaks if their values are greater than the minimum level. Previously there was always a slight green line at the bottom which I did not see due to fractional scaling.

Here's how it looks:

![FadersBranchRoundLevelsInset](https://github.com/LMMS/lmms/assets/9293269/ef3ee89e-3325-4f00-8b1e-bb374ab92a7d)

michaelgregorius commented:
The failing MSVC builds seem to be a problem with Cloudflare, see [here](https://github.com/actions/runner-images/issues/9476#issuecomment-1987316858).

michaelgregorius committed:
Merge remote-tracking branch 'upstream/master' into FaderInCodeWithGradients

Conflicts:
* plugins/Eq/EqEffect.cpp

michaelgregorius commented:
I have merged the current master to fix the self-inflicted merge conflict that was introduced with my pull request #7141. :sweat_smile:

Ready for review. :wink:

RoxasKH commented:
> Thanks for your feedback, @RoxasKH!
> 
> Most of your points should have been addressed by commit [b4bfc71](https://github.com/LMMS/lmms/commit/b4bfc7174fa6f551e85fd7ecf2fd650157588c74):
> 
>     * It introduces a margin in all directions between the level borders and the level meters.
> 
>     * Levels are now also rendered as rounded rectangles similar to the level borders.
> 
>     * The radius of the rounded rectangles was increased to make it more pronounced.
> 
>     * The margins have been decreased so that the level readings become wider.
> 
>     * The unity indicator is now rendered with the same width as the level meters so that they become fully covered by them.
> 
> 
> Other improvements:
> 
>     * Only render the levels and peaks if their values are greater than the minimum level. Previously there was always a slight green line at the bottom which I did not see due to fractional scaling.
> 
> 
> Here's how it looks:

Looks good to me.

We're just losing the small border highlight around them, but i guess it's not a big deal, you choose if you feel like it's worth being perfectionist and working on that or not.

I have some other concerns about the mixer stuff placement, positioning and width (and cursor not changing to pointer when hovering the send button), but i'm sure those were merged with another PR as i saw the same things happening in the current nightly build.

It's stuff that can be fixed later somewhere else and doesn't need to be addressed here tho.

Veratil commented on plugins/Eq/EqEffect.cpp:
```diff
@@ -313,41 +312,41 @@ float EqEffect::peakBand( float minF, float maxF, EqAnalyser *fft, int sr )
 {
 	m_eqControls.m_lowShelfPeakR = m_eqControls.m_lowShelfPeakL =
-			peakBand( m_eqControls.m_lowShelfFreqModel.value()
+			linearPeakBand( m_eqControls.m_lowShelfFreqModel.value()
```
```suggestion
			linearPeakBand(m_eqControls.m_lowShelfFreqModel.value()
```
Along with the others below

michaelgregorius replied:
Fixed with commit f2d63ecca1b which also introduced a lambda to reduce code repetition.

Veratil commented on include/lmms_math.h:
```diff
@@ -352,6 +353,28 @@ static inline int numDigitsAsInt(float f)
 	return digits;
 }
 
+template <typename T>
```
Do we want to limit `T` to certain types, or just let the compiler error if we try to use a type that isn't defined with any mathematical operators?

michaelgregorius replied:
I'd prefer the compiler error because that way the class gives more flexibility.

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -61,75 +62,50 @@ namespace lmms::gui
-	m_fMaxPeak( 1.1 ),
-	m_back(embed::getIconPixmap("fader_background")),
-	m_leds(embed::getIconPixmap("fader_leds")),
+Fader::Fader(FloatModel* model, const QString& name, QWidget* parent) :
```
If we set the defaults in `Fader.h` instead then we don't need to have this huge initialization list.

michaelgregorius replied:
Having just read [this discussion](https://stackoverflow.com/questions/28413154/c-initializing-variables-in-header-vs-with-constructor) about the pros and cons I think I'd rather keep it in the implementation file.

Veratil replied:
Can you explain more of your opinion?

michaelgregorius replied:
Mostly the argument on having to potentially recompile lots of sources when a value is changed. It somehow feels like the default values should be an implementation detail.

On the other hand, if the values were defined in the header files, then looking at the header would suffice for a developer to know, if the default values are as expected or whether an instance has to be potentially reconfigured after construction.

I think there is no 100% correct way.

Veratil replied:
> Mostly the argument on having to potentially recompile lots of sources when a value is changed. It somehow feels like the default values should be an implementation detail.

How often are we changing the default values? Also, giving the defaults in the class is part of the [core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c45-dont-define-a-default-constructor-that-only-initializes-data-members-use-default-member-initializers-instead) not once [but twice](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c48-prefer-default-member-initializers-to-member-initializers-in-constructors-for-constant-initializers).

michaelgregorius replied:
Ok, you have convinced me! :sweat_smile: Adjusted with commit ad8a31daa37.

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -61,75 +62,50 @@ namespace lmms::gui
 	}
 
-	init(_model, _name);
+	setWindowTitle( name );
+	setAttribute( Qt::WA_OpaquePaintEvent, false );
```
```suggestion
	setWindowTitle(name);
	setAttribute(Qt::WA_OpaquePaintEvent, false);
```

michaelgregorius replied:
Fixed with commit 2bec339fb8a.

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -61,75 +62,50 @@ namespace lmms::gui
+	setMinimumSize(minimumSize);
+	resize(minimumSize);
+	setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::MinimumExpanding);
+	setModel( model );
+	setHintText( "Volume:","%");
```
```suggestion
	setModel(model);
	setHintText("Volume:", "%");
```

michaelgregorius replied:
Fixed with commit 2bec339fb8a.

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -61,75 +62,50 @@ namespace lmms::gui
-	}
-
-	init(model, name);
+	m_knob = knob;
```
Couldn't this just be in the initialization list?

michaelgregorius replied:
This does not work because this constructor first calls the other one which initializes the knob to the default pixmap "fader_knob". This call then overrides the knob pixmap.

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -315,154 +284,207 @@ void Fader::updateTextFloat()
 }
 
-void Fader::paintDBFSLevels(QPaintEvent * ev, QPainter & painter)
+void Fader::paintLevels(QPaintEvent * ev, QPainter & painter, bool linear)
```
```suggestion
void Fader::paintLevels(QPaintEvent* ev, QPainter& painter, bool linear)
```

michaelgregorius replied:
Fixed with commit 2bec339fb8a.

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -315,154 +284,207 @@ void Fader::updateTextFloat()
-	int height = m_back.height();
-	int width = m_back.width() / 2;
-	int center = m_back.width() - width;
+	std::function<float(float value)> mapper = [this](float value) { return ampToDbfs(qMax<float>(0.0001, value)); };
```
Why not `auto mapper = ...`?
You can see I did this in the [Piano Roll](https://github.com/LMMS/lmms/blob/master/src/gui/editors/PianoRoll.cpp#L3111).

Veratil replied:
Gotcha, then the only thing I have to say is you don't need `value` in `std::function<float(float value)>`. It should just be `std::function<float(float)>`

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -315,154 +284,207 @@ void Fader::updateTextFloat()
 
-	float const maxDB(ampToDbfs(m_fMaxPeak));
-	float const minDB(ampToDbfs(m_fMinPeak));
+	float const mappedMinPeak(mapper(m_fMinPeak));
```
With most of the variable types here, I'd rather `const` be on the left (`const float`) to be more in line with the majority of the codebase.

michaelgregorius replied:
Done with commit b3a1dd9e69d although I don't like this ambiguity in C++. If you always write them to the right then you can simply read statements from right to left to get an exact idea of what something is.
```
int const * foo;
```
`foo` is a (non-const) pointer to a constant integer value.
```
int * const bar;
```
`bar` is a const pointer to a non-const integer value.
```
int const * const baz;
```
`baz` is a const pointer to a const integer value.

michaelgregorius replied:
The C++ standard or the coding guidelines of LMMS? :wink:

Veratil replied:
Our coding conventions of course. The standard doesn't define which order I'm pretty sure. 😂

michaelgregorius replied:
The standard usually goes along the line of "everybody can do however they please and here are 100 options". 😂

Would be nice to change this in LMMS' coding guidelines but I guess it would be a lot of work to adjust this everywhere in the code.

Veratil replied:
It would be a lot of work, yes. We'd also have to put it up to a vote.

michaelgregorius replied:
I guess it might also potentially break several existing pull requests. So it's a bit of a question if it is really worth it (even though I like "const to the right" :wink:).

Veratil commented on src/gui/widgets/Fader.cpp:
```diff
@@ -315,154 +284,207 @@ void Fader::updateTextFloat()
+	int const distanceBetweenMeters = 2;
+
+	int const numberOfMeters = 2;
+	int const meterWidth = (baseRect.width() - 2 * margin - distanceBetweenMeters * (numberOfMeters - 1)) / numberOfMeters;
```
I feel we should be explicit with parentheses in math like this that we do in fact mean `(2 * margin)` and `(distanceBetweenMeters * ...)`. Just a thought.

michaelgregorius replied:
Fixed with commit 2bec339fb8a. It introduces intermediate variables to make the intentions a bit clearer. This also reduces the number of parentheses.

Veratil commented on include/Fader.h:
```diff
@@ -66,17 +66,17 @@ class LMMS_EXPORT Fader : public QWidget, public FloatModelView
-
-	Fader( FloatModel * _model, const QString & _name, QWidget * _parent );
-	Fader( FloatModel * _model, const QString & _name, QWidget * _parent, QPixmap * back, QPixmap * leds, QPixmap * knob );
+	Q_PROPERTY(QColor peakOk READ peakOk WRITE setPeakOk)
+	Q_PROPERTY(QColor peakClip READ peakClip WRITE setPeakClip)
+	Q_PROPERTY(QColor peakWarn READ peakWarn WRITE setPeakWarn)
```
Would be useful to test, but if it's too much trouble to figure out don't worry about it.

michaelgregorius replied:
Done with commit 1bd81d884cf for all properties that are not set explicitly from outside of the class.

michaelgregorius committed:
Reduce code repetition

Reduce code repetition in `EqEffect::setBandPeaks` by introducing a lambda. Adjust code formatting.

michaelgregorius committed:
Code review changes

Code review changes in `Fader.cpp`. Mostly whitespace adjustments.

Split up the calculation of the meter width to make it more understandable. This also reduces the number of parentheses.

michaelgregorius committed:
Move const to the left of types

michaelgregorius committed:
Use MEMBER instead of READ/WRITE

Use `MEMBER` instead of `READ`/`WRITE` for some properties that are not called explicitly from outside of the class.

michaelgregorius committed:
Use default member initializers for Fader

Use default member initializers for the members in `Fader` that have previously been initialized in the constructor list.

Mark `ampToDbfs` and `dbfsToAmp` as `constexpr`. The function `dbfsToAmp` is used in the initialzation and this way nothing has to be computed at runtime.

michaelgregorius committed:
Remove constexpr again

Remove `constexpr` from `ampToDbfs` and ` dbfsToAmp` again because the MSVC compiler does not like it. Seems that for some reason `log10f` and `std::pow` are not `constexpr` there (at least in the version used on the build server).

Veratil commented on include/Fader.h:
```diff
@@ -122,10 +114,7 @@ class LMMS_EXPORT Fader : public QWidget, public FloatModelView
-
-	void paintDBFSLevels(QPaintEvent *ev, QPainter & painter);
-	void paintLinearLevels(QPaintEvent *ev, QPainter & painter);
+	void paintLevels(QPaintEvent *ev, QPainter & painter, bool linear = false);
```
```suggestion
	void paintLevels(QPaintEvent* ev, QPainter& painter, bool linear = false);
```

Veratil commented on plugins/Eq/EqEffect.cpp:
```diff
@@ -287,70 +287,61 @@ bool EqEffect::processAudioBuffer( sampleFrame *buf, const fpp_t frames )
 
 void EqEffect::setBandPeaks( EqAnalyser *fft, int samplerate )
 {
+	auto computePeakBand = [this, fft, samplerate](FloatModel const & freqModel, FloatModel const & bwModel)
```
I'm fairly certain we can do this with just `&`:
```suggestion
	auto computePeakBand = [&](const FloatModel& freqModel, const FloatModel& bwModel)
```

michaelgregorius replied:
You're right. Fixed with commit bb212eebbd3.

Veratil commented on plugins/Eq/EqEffect.cpp:
```diff
@@ -287,70 +287,61 @@ bool EqEffect::processAudioBuffer( sampleFrame *buf, const fpp_t frames )
 	m_eqControls.m_lowShelfPeakR = m_eqControls.m_lowShelfPeakL =
-			peakBand( m_eqControls.m_lowShelfFreqModel.value()
-					  * ( 1 - m_eqControls.m_lowShelfResModel.value() * 0.5 ),
+			linearPeakBand(m_eqControls.m_lowShelfFreqModel.value()
```
This is tabbed twice, we only need once. Similar with the rest of the argument lines below it.

michaelgregorius replied:
Fixed with commit bb212eebbd3.

Veratil commented on plugins/Eq/EqEffect.cpp:
```diff
@@ -287,70 +287,61 @@ bool EqEffect::processAudioBuffer( sampleFrame *buf, const fpp_t frames )
 
 	m_eqControls.m_highShelfPeakL = m_eqControls.m_highShelfPeakR =
-			peakBand( m_eqControls.m_highShelfFreqModel.value(),
+			linearPeakBand(m_eqControls.m_highShelfFreqModel.value(),
```
Tabbed twice and below

michaelgregorius committed:
Comprehensive code formatting in Fader class

michaelgregorius committed:
Improved capture and formatting

michaelgregorius committed:
Merge remote-tracking branch 'upstream/master' into FaderInCodeWithGradients

michaelgregorius committed:
Make code clearer

Make code clearer in `Fader::FadermouseDoubleClickEvent`. Only divide if the dialog was accepted with OK.
